### PR TITLE
#113: Added Spring Security Authorization

### PR DESCRIPTION
**Pull request by Rohit90700**
This PR is fixing issue #113 and is an extension to PR
 https://github.com/devonfw/my-thai-star/pull/221 
 https://github.com/devonfw/my-thai-star/pull/196
(roles and permission based Authorization) with existing MTS.

### CODE REVIEW

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/clustermanagement/logic/impl/ClustermanagementImpl.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public ClustermanagementImpl() {
   }
 
   @Override
-  @RolesAllowed(Roles.MANAGER)
+  @RolesAllowed({ ApplicationAccessControlConfig.PERMISSION_GET_GEO_CLUSTER })
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/clustermanagement/logic/impl/ClustermanagementImpl.java]: Done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/common/base/TokenAuthenticationService.java [hunk 2]
```diff
@@ -98,8 +100,8 @@ static Authentication getAuthentication(HttpServletRequest request) {
     if (token != null) {
 
       // The JWT parser will throw an exception if the token is not well formed or the token has expired
-      String user =
-          Jwts.parser().setSigningKey(SECRET).parseClaimsJws(token.replace(TOKEN_PREFIX, "")).getBody().getSubject();
+      String user = Jwts.parser().setSigningKey(SECRET).parseClaimsJws(token.replace(TOKEN_PREFIX, "")).getBody()
+          .getSubject();
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/common/base/TokenAuthenticationService.java]: created issue for it - https://github.com/devonfw/my-thai-star/issues/282

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/common/impl/security/ApplicationAccessControlConfig.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  public static final String PERMISSION_FIND_USER_ROLE = PREFIX + "FindUserRole";
+
+  public static final String PERMISSION_DELETE_USER_ROLE = PREFIX + "DeleteUserRole";
+
+  public static final String PERMISSION_SAVE_USER_ROLE = PREFIX + "SaveUserRole";
+
+  public static final String PERMISSION_GET_GEO_CLUSTER = PREFIX + "GetGeoCluster";
+
+  public static final String PERMISSION_GET_NEXT_WEEK_PREDICTION = PREFIX + "GetNextWeekPrediction";
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/common/impl/security/ApplicationAccessControlConfig.java]: done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java [hunk 4]
```diff
@@ -17,16 +19,40 @@
 @Component
 public class UsermanagementDummyImpl extends AbstractBeanMapperSupport implements Usermanagement {
 
+  @Inject
+  private com.devonfw.application.mtsj.usermanagement.logic.api.Usermanagement usermanagementcore;
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java]: Done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java [hunk 5]
```diff
@@ -17,16 +19,40 @@
...
 
+    UserEto userEto = this.usermanagementcore.findUserbyName(login);
     // this is only a dummy - please replace with a real implementation
     UserDetailsClientTo profile = new UserDetailsClientTo();
+    profile.setId(userEto.getId());
     profile.setName(login);
     profile.setFirstName("Peter");
     profile.setLastName(login);
-    profile.setRole(Role.WAITER);
+    profile.setRole(setRole(userEto.getUserRoleId()));
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java]: Done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java [hunk 6]
```diff
@@ -17,16 +19,40 @@
...
+   */
+  private Role setRole(long roleId) {
+
+    switch (String.valueOf(roleId).trim()) {
+      case "1":
+        return Role.WAITER;
+      case "0":
+        return Role.CUSTOMER;
+      case "2":
+        return Role.MANAGER;
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/general/logic/impl/UsermanagementDummyImpl.java]: Done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/ordermanagement/logic/impl/OrdermanagementImpl.java [hunk 7]
```diff
@@ -130,7 +130,7 @@ public OrderCto findOrder(Long id) {
   }
 
   @Override
-  @RolesAllowed({ Roles.WAITER, Roles.MANAGER })
+  @RolesAllowed({ ApplicationAccessControlConfig.PERMISSION_FIND_ORDER })
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/ordermanagement/logic/impl/OrdermanagementImpl.java]: Done

File: java/mtsj/core/src/main/java/com/devonfw/application/mtsj/usermanagement/logic/impl/UsermanagementImpl.java [hunk 8]
```diff
@@ -31,165 +31,174 @@
...
-        super();
-    }
-
-    @Override
-    public UserEto findUser(Long id) {
-
-        LOG.debug("Get User with id {} from database.", id);
-        return getBeanMapper().map(getUserDao().find(id), UserEto.class);
+  /**
+   * Logger instance.
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/main/java/com/devonfw/application/mtsj/usermanagement/logic/impl/UsermanagementImpl.java]: Done

File: java/mtsj/core/src/test/java/com/devonfw/application/mtsj/dishmanagement/logic/impl/DishmanagementTest.java [hunk 9]
```diff
@@ -53,14 +53,14 @@ public void filterDishes() {
...
-    PageRequest pageable = PageRequest.of(0, 100, new Sort(Direction.DESC, "price"));
+    criteria.setSearchBy("Thai Spicy Basil Fried Rice");
+    PageRequest pageable = PageRequest.of(0, 100, new Sort(Direction.DESC, "id"));
     criteria.setPageable(pageable);
     Page<DishCto> result = this.dishmanagement.findDishCtos(criteria);
 
     assertThat(result).isNotNull();
     assertThat(result.getContent().size()).isGreaterThan(0);
-    assertThat(result.getContent().get(0).getDish().getId()).isEqualTo(1L);
+    assertThat(result.getContent().get(0).getDish().getId()).isEqualTo(0L);
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/test/java/com/devonfw/application/mtsj/dishmanagement/logic/impl/DishmanagementTest.java]: Done

File: java/mtsj/core/src/test/java/com/devonfw/application/mtsj/general/common/TestUtil.java [hunk 10]
```diff
@@ -16,7 +26,18 @@
...
+    Set<AccessControl> accessControlSet = new HashSet<>();
+    for (String id : permissions) {
+      boolean success = accessControlConfig.collectAccessControls(id, accessControlSet);
+      if (!success) {
+      }
+    }
+    for (AccessControl accessControl : accessControlSet) {
+      authorities.add(new AccessControlGrantedAuthority(accessControl));
+    }
+    Authentication authentication = new TestingAuthenticationToken(login, login, authorities);
```
[the remark this answers is not in the document]
Comment (Rohit90700) [reply java/mtsj/core/src/test/java/com/devonfw/application/mtsj/general/common/TestUtil.java]: Need to create a separate issue for it.

**Comment by Rohit90700** [comment]
Hi @hohwille , Changes are done ,Please review once again and merge it if no more changes required.

File: java/mtsj/core/src/test/java/com/devonfw/application/mtsj/general/common/TestUtil.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (hohwille) [reply java/mtsj/core/src/test/java/com/devonfw/application/mtsj/general/common/TestUtil.java]: Created issue #284

**Pull request merged**